teps/0011-redirecting-step-output-streams.md

```diff
@@ -0,0 +1,212 @@
+
+* Allow tool developers to create CI pipeline tooling that can use Task results on *any* image specified by the end users without the above limitations. As Task authors, one can choose which images to use to work around the limitations. But as a tool developer, one cannot or should not control what images their end users can use, and cannot assume any details about the images.
+
+### Goals
+
+1. Allow the stdout/stderr of a step to be consumed by another step.
+1. Enable a user to configure the path where the output streams are written.
+
+### Non-Goals
+
+1. Parse stdout/stderr of a step into a structured format (e.g., JSON) and extract information from certain fields.
+
+## Requirements
+
+* Add new fields for a step to specify paths to redirect stdout/stderr to.
+* Users should be able to observe the output streams through Pod logs even if `StdoutPath` or `StderrPath` is specified. In other words, output streams should be duplicated instead of simply redirected.
+* Clearly documents the restrictions when `StdoutPath` and `StderrPath` are set to Task result paths, and encourage Task authors to set the paths to workspace paths, especially if they want to exchange large amount of data between Tasks.
+* Provide examples to use step redirection for the use cases mentioned in [Motivation](#motivation).
+
+## Proposal
+
+### Option 1: Allowing Users to Specify Redirection Paths
+
+The following new fields will be added to the `Step` struct:
+
+```go
+type Step struct {
+    ...
+    // Path to duplicate stdout stream to on container's local filesystem.
+    // +optional
+    StdoutPath string `json:"stdoutPath,omitempty"`
+    // Path to duplicate stderr stream to on container's local filesystem.
+    // +optional
+    StderrPath string `json:"stderrPath,omitempty"`
+}
+```
+
+Once `StdoutPath` or `StderrPath` is specified, the corresponding output stream will be duplicated to both the given file and the standard output stream of the container, so users can still view the output through the Pod log API. If both `StdoutPath` and `StderrPath` are set to the same value, the two streams will be written to two file descriptors referring to the same file description, so outputs from both stream will be interleaved in the same file, but there will be no ordering guarantee on the data.  If multiple steps' `StdoutPath` fields are set to the same value, the file content will be overwritten by the last outputting step.
+
+Variable substitution will be applied to the new fields, so one could specify `$(results.<name>.path)` to the `StdoutPath` field to extract the stdout of a step into a Task result. No new variable substitution for accessing the values of `StdoutPath` and `StderrPath` fields will be provided so variable substitution can remain single-pass.
```

Comment:
> `$(results.<name>.path)` to the `StdoutPath` field to extract the stdout of a step into a Task result

I'm a bit confused by this, you've got an example below: `stdout: $(results.project-id.path)`

in this case, what path does stdout actually get written to? is this magically writing stdout to a location on disk and then also exposing it as a result?

i dont think we should automatically wire stdout/stderr to any results. if we do, i think we could do it in a subsequent proposal after this

my alternative suggestion for the example below is that if you want to expose stdout as a result, you'd need to add a step after `parse-project-id` that writes stdout as a result

> No new variable substitution for accessing the values of `StdoutPath` and `StderrPath` fields will be provided so variable substitution can remain single-pass.

can you explain a bit more? i think it would be very convenient to use `$(steps.foo.stdout.path)` or similar, and that's exactly how id want to use this feature, i wouldnt want to have to hardcode the path in multiple places

Reply:
> `$(results.project-io.path)` would be interpolated and be the path of the results (so `/tekton/result/project-io` or something), and thus, stdout would be written in. Just like if you use `$(workspaces.some.path)/foo`.

oh that makes sense! i didnt realize you could do that and i've been hardcoding the path everywhere. just ignore that comment from me then @chhsia0!

> So one of the problem with `$(steps.foo.stdout.path)` is that, it may be used without the previous task having `stdout` captured. You don't necessarly hardcode the path, or at least the full path ; most likely you would use `$(workspace.foo.path)/somewhere/afile` as a path, and re-use the same later on.

coud you explain a bit more about this? we could check if folks are using `$(steps.foo.stdout.path)` for a step that doesn't have stdout capturing configured. it still seems really useful to me

Reply:
> we could check if folks are using $(steps.foo.stdout.path) for a step that doesn't have stdout capturing configured

actually without this feature, we wouldn't be able to help out anyone making this mistake

Reply:
This is exactly Option 1 vs Option 2 if we also want the variable substitution mechanism to remain single-pass. With the single-pass assumption, Option 1 gives us a way to capture stdout into results, workspaces, volumes, ... but it's harder to validate if the specified path exists (and IMO it's Task authors' responsibility: they're responsible to ensure any path used in their Tasks is valid); Option 2 provides more guarantees and validations, but is more limited.